cilium/bpf_metadata.cc

```diff
@@ -552,13 +552,15 @@ Network::FilterStatus Instance::onAccept(Network::ListenerFilterCallbacks& cb) {
 
     // Make Cilium Policy data available to filters and upstream connection (Cilium TLS Wrapper) as
     // filter state.
+    auto policy_fs = socket_metadata->buildCiliumPolicyFilterState();
     cb.filterState().setData(
-        Cilium::CiliumPolicyFilterState::key(), socket_metadata->buildCiliumPolicyFilterState(),
+        Cilium::CiliumPolicyFilterState::key(), policy_fs,
         StreamInfo::FilterState::StateType::ReadOnly, StreamInfo::FilterState::LifeSpan::Connection,
         StreamInfo::StreamSharingMayImpactPooling::SharedWithUpstreamConnection);
 
     // Restoring original source address on the upstream socket
-    socket_options->push_back(socket_metadata->buildSourceAddressSocketOption(config_->so_linger_));
+    socket_options->push_back(
+        socket_metadata->buildSourceAddressSocketOption(config_->so_linger_, policy_fs));
```

Comment:
is it possible to avoid passing the policy filterstate to the socket option and instead looking it up once the socket option logic gets executed?

Reply:
iiuc, socket option is pretty low level, while filterstate is in the connection level, so I don't know how.

Reply:
the above was considered, however, imo using filter state is cleaner, especially after your comment on dedicated filter state.

Reply:
IMO this is now resolved with the dedicated filter state object.